sale_invoice_plan/models/sale_invoice_plan.py

```diff
@@ -0,0 +1,227 @@
+            self.env["ir.config_parameter"]
+            .sudo()
+            .get_param("sale.default_deposit_product_id")
+        )
+        if deposit_product_id:
+            lines = invoices.mapped("invoice_line_ids").filtered(
+                lambda line: line.product_id.id != int(deposit_product_id)
+            )
+            amount_invoiced = sum(lines.mapped("price_subtotal"))
+        return amount_invoiced
+
+    @api.depends("invoice_move_ids.state")
+    def _compute_invoiced(self):
+        for rec in self:
+            invoiced = rec.invoice_move_ids.filtered(
+                lambda line: line.state in ("draft", "posted")
+            )
+            rec.invoiced = True if invoiced else False
+            rec.amount_invoiced = (
+                sum(invoiced.mapped("amount_untaxed"))
+                if rec.invoice_type == "advance"
+                else rec._get_amount_invoice(invoiced[:1])
+            )
+
+    def _compute_last(self):
+        for rec in self:
+            last = max(rec.sale_id.invoice_plan_ids.mapped("installment"))
+            rec.last = rec.installment == last
+
+    def _compute_new_invoice_quantity(self, invoice_move):
+        self.ensure_one()
+        if self.last:  # For last install, let the system do the calc.
+            return
+        percent = self.percent
+        move = invoice_move.with_context(check_move_validity=False)
+        for line in move.invoice_line_ids:
+            self._update_new_quantity(line, percent)
+        move.line_ids.filtered(
+            lambda x: x.display_type
+            not in ("product", "line_section", "line_note", "payment_term", "tax")
+        ).unlink()
+
+    def _update_new_quantity(self, line, percent):
+        """Hook function"""
+        if not len(line.sale_line_ids) >= 0:
+            raise UserError(_("No matched order line for invoice line"))
+        order_line = fields.first(line.sale_line_ids)
+        if order_line.is_downpayment:  # based on 1 unit
+            line.write({"quantity": -percent / 100})
+        else:
+            plan_qty = self._get_plan_qty(order_line, percent)
+            prec = order_line.product_uom.rounding
+            if plan_qty:
+                plan_qty = float_round(plan_qty, precision_rounding=prec)
+            if float_compare(abs(plan_qty), abs(line.quantity), prec) == 1:
+                raise ValidationError(
+                    _(
+                        "Plan quantity: %(plan_qty)s, exceed invoiceable quantity: "
+                        "%(invoiceable_qty)s"
+                        "\nProduct should be delivered before invoice"
+                    )
+                    % {"plan_qty": plan_qty, "invoiceable_qty": line.quantity}
+                )
+            line.write({"quantity": plan_qty})
+
+    @api.model
+    def _get_plan_qty(self, order_line, percent):
+        plan_qty = order_line.product_uom_qty * (percent / 100)
+        return plan_qty
+
+    def unlink(self):
```

Comment:
An `@api.ondelete` method could be used instead (check https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0)

Reply:
hello! Thanks for the feedback. I implemented the ondelete  method as you suggested.